superset-frontend/src/dashboard/util/isValidChild.ts

```diff
@@ -105,12 +105,36 @@ const parentMaxDepthLookup = {
   [MARKDOWN_TYPE]: {},
 };
 
-export default function isValidChild({ parentType, childType, parentDepth }) {
+type ParentMaxDepthLookup = typeof parentMaxDepthLookup;
+
+interface IsValidChildProps {
+  parentType?: unknown;
+  childType?: unknown;
+  parentDepth?: unknown;
+}
+
+interface ValidChild<
+  P extends keyof ParentMaxDepthLookup,
+  C extends keyof ParentMaxDepthLookup[P],
+  D extends ParentMaxDepthLookup[P][C]
+> {
+  parentType: P;
+  childType: C;
+  parentDepth: D;
+}
+
+export default function isValidChild<
+  P extends keyof ParentMaxDepthLookup = any,
+  C extends keyof ParentMaxDepthLookup[P] = any,
+  D extends ParentMaxDepthLookup[P][C] = any
+>(child: IsValidChildProps): child is ValidChild<P, C, D> {
```

Comment:
The idea here is to offer a version which doesn't need the generics (falling back to any).
but if generics are provided, it can apply a predicate to narrow the type https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards